// @testmuai/browser-cloud/dist/testmu-cloud/services/captcha-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CaptchaService = void 0;
/**
 * CaptchaService - Captcha Solving Integration
 *
 * Supports integration with external services like 2Captcha, Anti-Captcha,
 * or LambdaTest's built-in solving.
 *
 * Configure via environment variables:
 * - CAPTCHA_API_KEY: API key for the captcha service
 * - CAPTCHA_SERVICE: Service to use ('2captcha', 'anticaptcha', 'lambdatest')
 */
class CaptchaService {
    constructor() {
        this.jobs = new Map();
        this.apiKey = process.env.CAPTCHA_API_KEY || '';
        this.service = process.env.CAPTCHA_SERVICE || '2captcha';
    }
    async solveImage(sessionId, params) {
        const jobId = `captcha_${sessionId}_${Date.now()}`;
        // Initialize job status
        this.jobs.set(jobId, {
            id: jobId,
            status: 'pending'
        });
        // Start async solving
        this.processCapture(jobId, params);
        return {
            id: jobId,
            status: 'pending'
        };
    }
    async status(sessionId) {
        // Find the most recent job for this session
        const jobs = Array.from(this.jobs.entries())
            .filter(([id]) => id.includes(sessionId))
            .sort((a, b) => {
            const timeA = parseInt(a[0].split('_').pop() || '0');
            const timeB = parseInt(b[0].split('_').pop() || '0');
            return timeB - timeA;
        });
        if (jobs.length === 0) {
            return {
                id: '',
                status: 'failed',
                error: 'No captcha job found for session'
            };
        }
        return jobs[0][1];
    }
    /**
     * Get status by job ID
     */
    async getJobStatus(jobId) {
        return this.jobs.get(jobId) || {
            id: jobId,
            status: 'failed',
            error: 'Job not found'
        };
    }
    /**
     * Process captcha (simulated or real API call)
     */
    async processCapture(jobId, params) {
        const startTime = Date.now();
        try {
            // Update to processing
            this.jobs.set(jobId, {
                id: jobId,
                status: 'processing'
            });
            let solution;
            if (this.apiKey) {
                // Real API integration
                solution = await this.callExternalService(params);
            }
            else {
                // Simulated solving for development
                console.log(`[CaptchaService] No API key configured, using simulated solving`);
                await this.simulateDelay(2000);
                solution = `simulated_token_${Date.now()}`;
            }
            // Update to solved
            this.jobs.set(jobId, {
                id: jobId,
                status: 'solved',
                solution,
                solvingTime: Date.now() - startTime
            });
        }
        catch (error) {
            this.jobs.set(jobId, {
                id: jobId,
                status: 'failed',
                error: error instanceof Error ? error.message : String(error),
                solvingTime: Date.now() - startTime
            });
        }
    }
    /**
     * Call external captcha solving service
     */
    async callExternalService(params) {
        switch (this.service) {
            case '2captcha':
                return this.solve2Captcha(params);
            case 'anticaptcha':
                return this.solveAntiCaptcha(params);
            case 'lambdatest':
                return this.solveLambdaTest(params);
            default:
                throw new Error(`Unknown captcha service: ${this.service}`);
        }
    }
    /**
     * 2Captcha API integration (placeholder)
     */
    async solve2Captcha(params) {
        // TODO: Implement actual 2Captcha API calls
        // POST to https://2captcha.com/in.php
        // Poll https://2captcha.com/res.php
        console.log('[CaptchaService] 2Captcha integration - implement API calls');
        throw new Error('2Captcha integration not yet implemented. Set CAPTCHA_API_KEY and implement API calls.');
    }
    /**
     * Anti-Captcha API integration (placeholder)
     */
    async solveAntiCaptcha(params) {
        // TODO: Implement actual Anti-Captcha API calls
        console.log('[CaptchaService] Anti-Captcha integration - implement API calls');
        throw new Error('Anti-Captcha integration not yet implemented.');
    }
    /**
     * TestMu AI Browser Cloud captcha solving (placeholder)
     */
    async solveLambdaTest(params) {
        // TODO: Implement TestMu AI's captcha solving API
        console.log('[CaptchaService] TestMu AI captcha solving - implement API calls');
        throw new Error('TestMu AI captcha integration not yet implemented.');
    }
    /**
     * Simulate delay for development/testing
     */
    simulateDelay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
    // ================== Legacy API for backwards compatibility ==================
    /**
     * @deprecated Use solveImage() instead
     */
    async solve(params) {
        const response = await this.solveImage('legacy', params);
        // Wait for solving to complete (for backwards compatibility)
        await this.simulateDelay(2500);
        const status = await this.getJobStatus(response.id);
        return {
            id: status.id,
            status: status.status === 'solved' ? 'SOLVED' : status.status.toUpperCase(),
            code: status.solution
        };
    }
    /**
     * @deprecated Use getJobStatus() instead
     */
    async getResult(id) {
        const status = await this.getJobStatus(id);
        return {
            id: status.id,
            status: status.status === 'solved' ? 'SOLVED' : status.status.toUpperCase(),
            code: status.solution
        };
    }
}
exports.CaptchaService = CaptchaService;
