// @testmuai/browser-cloud/dist/testmu-cloud/services/computer-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComputerService = void 0;
/**
 * ComputerService - AI Agent Mouse/Keyboard Control
 *
 * Provides programmatic control for AI agents to interact with browser sessions
 * using mouse movements, clicks, keyboard input, and screenshots.
 *
 * This is a critical feature for AI agent use cases (like Claude computer use).
 */
class ComputerService {
    /**
     * Execute a computer action on a page
     */
    async execute(page, params) {
        try {
            let output = '';
            let base64_image;
            switch (params.action) {
                case 'move':
                    if (params.coordinate) {
                        await page.mouse.move(params.coordinate[0], params.coordinate[1]);
                        output = `Moved mouse to (${params.coordinate[0]}, ${params.coordinate[1]})`;
                    }
                    break;
                case 'click':
                    if (params.coordinate) {
                        await page.mouse.click(params.coordinate[0], params.coordinate[1]);
                        output = `Clicked at (${params.coordinate[0]}, ${params.coordinate[1]})`;
                    }
                    break;
                case 'double_click':
                    if (params.coordinate) {
                        await page.mouse.click(params.coordinate[0], params.coordinate[1], { clickCount: 2 });
                        output = `Double-clicked at (${params.coordinate[0]}, ${params.coordinate[1]})`;
                    }
                    break;
                case 'right_click':
                    if (params.coordinate) {
                        await page.mouse.click(params.coordinate[0], params.coordinate[1], { button: 'right' });
                        output = `Right-clicked at (${params.coordinate[0]}, ${params.coordinate[1]})`;
                    }
                    break;
                case 'scroll':
                    const deltaX = params.deltaX || 0;
                    const deltaY = params.deltaY || 0;
                    // Move to coordinate first if provided
                    if (params.coordinate) {
                        await page.mouse.move(params.coordinate[0], params.coordinate[1]);
                    }
                    await page.mouse.wheel({ deltaX, deltaY });
                    output = `Scrolled by (${deltaX}, ${deltaY})`;
                    break;
                case 'type':
                    if (params.text) {
                        await page.keyboard.type(params.text);
                        output = `Typed: "${params.text}"`;
                    }
                    break;
                case 'key':
                    if (params.text) {
                        // Handle special keys like "Enter", "Escape", "Tab", etc.
                        await page.keyboard.press(params.text);
                        output = `Pressed key: ${params.text}`;
                    }
                    break;
                case 'screenshot':
                    const screenshot = await page.screenshot({
                        encoding: 'base64',
                        fullPage: false
                    });
                    base64_image = screenshot;
                    output = 'Screenshot captured';
                    break;
                default:
                    return {
                        error: `Unknown action: ${params.action}`,
                        system: 'computer-service'
                    };
            }
            // Take screenshot if requested with the action
            if (params.screenshot && params.action !== 'screenshot') {
                const screenshot = await page.screenshot({
                    encoding: 'base64',
                    fullPage: false
                });
                base64_image = screenshot;
            }
            return {
                output,
                base64_image,
                system: 'computer-service'
            };
        }
        catch (error) {
            return {
                error: error instanceof Error ? error.message : String(error),
                system: 'computer-service'
            };
        }
    }
    /**
     * Take a screenshot and return as base64
     */
    async screenshot(page, fullPage = false) {
        const screenshot = await page.screenshot({
            encoding: 'base64',
            fullPage
        });
        return screenshot;
    }
    /**
     * Move mouse to coordinates
     */
    async move(page, x, y) {
        await page.mouse.move(x, y);
    }
    /**
     * Click at coordinates
     */
    async click(page, x, y) {
        await page.mouse.click(x, y);
    }
    /**
     * Type text
     */
    async type(page, text) {
        await page.keyboard.type(text);
    }
    /**
     * Press a key
     */
    async press(page, key) {
        await page.keyboard.press(key);
    }
    /**
     * Scroll the page
     */
    async scroll(page, deltaX, deltaY) {
        await page.mouse.wheel({ deltaX, deltaY });
    }
}
exports.ComputerService = ComputerService;
