// @testmuai/browser-cloud/dist/testmu-cloud/services/credential-service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CredentialService = void 0;
const fs = __importStar(require("fs-extra"));
const path = __importStar(require("path"));
/**
 * CredentialService - Credential Management
 *
 * Manages stored credentials for automatic login functionality.
 * Credentials are matched by URL and can auto-submit login forms.
 *
 * Storage: .credentials/credentials.json
 *
 * Note: Credentials are stored in plain text. For production use,
 * implement encryption using a proper secret management solution.
 */
class CredentialService {
    constructor() {
        this.credentials = new Map();
        const credentialsDir = path.join(process.cwd(), '.credentials');
        fs.ensureDirSync(credentialsDir);
        this.credentialsFile = path.join(credentialsDir, 'credentials.json');
        this.loadCredentials();
    }
    /**
     * Create a new credential
     */
    async create(params) {
        const id = `cred_${Date.now()}`;
        const credential = {
            id,
            url: params.url,
            username: params.username,
            password: params.password,
            createdAt: new Date().toISOString()
        };
        this.credentials.set(id, credential);
        await this.saveCredentials();
        return credential;
    }
    /**
     * Update an existing credential
     */
    async update(id, params) {
        const existing = this.credentials.get(id);
        if (!existing) {
            throw new Error(`Credential not found: ${id}`);
        }
        const updated = {
            ...existing,
            ...params,
            id, // Prevent ID change
            updatedAt: new Date().toISOString()
        };
        this.credentials.set(id, updated);
        await this.saveCredentials();
        return updated;
    }
    /**
     * List credentials, optionally filtered by URL
     */
    async list(url) {
        const allCredentials = Array.from(this.credentials.values());
        if (url) {
            return allCredentials.filter(cred => this.urlMatches(cred.url, url));
        }
        // Return credentials without passwords for listing
        return allCredentials.map(cred => ({
            ...cred,
            password: '********'
        }));
    }
    /**
     * Get a specific credential by ID
     */
    async get(id) {
        return this.credentials.get(id) || null;
    }
    /**
     * Delete a credential
     */
    async delete(id) {
        this.credentials.delete(id);
        await this.saveCredentials();
    }
    /**
     * Delete all credentials
     */
    async deleteAll() {
        this.credentials.clear();
        await this.saveCredentials();
    }
    /**
     * Find credentials matching a URL
     */
    async findForUrl(url) {
        for (const credential of this.credentials.values()) {
            if (this.urlMatches(credential.url, url)) {
                return credential;
            }
        }
        return null;
    }
    /**
     * Check if any credentials exist for a URL
     */
    async hasCredentialsFor(url) {
        return (await this.findForUrl(url)) !== null;
    }
    // ================== Helper Methods ==================
    /**
     * Check if stored URL matches the target URL
     */
    urlMatches(storedUrl, targetUrl) {
        try {
            const stored = new URL(storedUrl);
            const target = new URL(targetUrl);
            // Match origin (protocol + host)
            return stored.origin === target.origin;
        }
        catch {
            // Fall back to simple string comparison
            return targetUrl.includes(storedUrl) || storedUrl.includes(targetUrl);
        }
    }
    /**
     * Load credentials from disk
     */
    async loadCredentials() {
        try {
            if (await fs.pathExists(this.credentialsFile)) {
                const data = await fs.readJson(this.credentialsFile);
                if (Array.isArray(data)) {
                    for (const cred of data) {
                        this.credentials.set(cred.id, cred);
                    }
                }
            }
        }
        catch (e) {
            console.error('[CredentialService] Error loading credentials:', e);
        }
    }
    /**
     * Save credentials to disk
     */
    async saveCredentials() {
        try {
            const data = Array.from(this.credentials.values());
            await fs.writeJson(this.credentialsFile, data, { spaces: 2 });
        }
        catch (e) {
            console.error('[CredentialService] Error saving credentials:', e);
        }
    }
}
exports.CredentialService = CredentialService;
