// @testmuai/browser-cloud/dist/testmu-cloud/services/local-browser-service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalBrowserService = void 0;
const http_1 = __importDefault(require("http"));
class LocalBrowserService {
    async launch() {
        console.log('Searching for local Chrome installation...');
        // Dynamic import for ESM-only chrome-launcher package
        const ChromeLauncher = await Promise.resolve().then(() => __importStar(require('chrome-launcher')));
        // Launch Chrome using chrome-launcher (automagically finds Chrome/Canary/Chromium)
        const chrome = await ChromeLauncher.launch({
            startingUrl: 'about:blank',
            chromeFlags: [
                '--headless=new', // Optional: could expose this in config
                '--disable-gpu',
                '--remote-debugging-port=0' // Random port
            ]
        });
        console.log(`Chrome launched on port ${chrome.port}`);
        // Fetch the WebSocket URL from the browser's DevTools protocol
        const versionData = await this.fetchVersion(chrome.port);
        const websocketUrl = versionData.webSocketDebuggerUrl;
        return {
            websocketUrl,
            kill: async () => {
                await chrome.kill();
            }
        };
    }
    fetchVersion(port) {
        return new Promise((resolve, reject) => {
            const req = http_1.default.get(`http://127.0.0.1:${port}/json/version`, (res) => {
                let data = '';
                res.on('data', (chunk) => data += chunk);
                res.on('end', () => {
                    try {
                        resolve(JSON.parse(data));
                    }
                    catch (e) {
                        reject(e);
                    }
                });
            });
            req.on('error', reject);
            req.end();
        });
    }
}
exports.LocalBrowserService = LocalBrowserService;
