// @testmuai/browser-cloud/dist/testmu-cloud/services/quick-actions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuickActionsService = void 0;
const puppeteer_extra_1 = __importDefault(require("puppeteer-extra"));
const puppeteer_extra_plugin_stealth_1 = __importDefault(require("puppeteer-extra-plugin-stealth"));
puppeteer_extra_1.default.use((0, puppeteer_extra_plugin_stealth_1.default)());
/**
 * QuickActionsService - TestMu AI Browser Cloud-compatible Quick Actions
 *
 * Provides quick operations for scraping, screenshots, and PDFs.
 * Can operate standalone or use existing sessions.
 */
class QuickActionsService {
    constructor() {
        this.sessionPages = new Map();
    }
    /**
     * Register a page for a session (for session-based quick actions)
     */
    registerSessionPage(sessionId, page) {
        this.sessionPages.set(sessionId, page);
    }
    /**
     * Unregister a session page
     */
    unregisterSessionPage(sessionId) {
        this.sessionPages.delete(sessionId);
    }
    /**
     * Get a page - either from existing session or create transient browser
     */
    async getPage(sessionId) {
        if (sessionId && this.sessionPages.has(sessionId)) {
            // Use existing session page
            const page = this.sessionPages.get(sessionId);
            return {
                page,
                cleanup: async () => { } // No cleanup needed for existing session
            };
        }
        // Get Chrome executable path using chrome-launcher (dynamic import for ESM)
        const Launcher = await Promise.resolve().then(() => __importStar(require('chrome-launcher')));
        const installations = Launcher.Launcher.getInstallations();
        const chromePath = installations.length > 0 ? installations[0] : undefined;
        if (!chromePath) {
            throw new Error('Chrome not found. Please install Chrome or use a session-based quick action.');
        }
        // Launch transient browser with Chrome path
        const browser = await puppeteer_extra_1.default.launch({
            headless: true,
            executablePath: chromePath
        });
        const page = await browser.newPage();
        return {
            page: page,
            cleanup: async () => {
                await browser.close();
            }
        };
    }
    /**
     * Wait for page to be ready
     */
    async waitForPage(page, delay, waitFor) {
        if (waitFor) {
            await page.waitForSelector(waitFor, { timeout: 30000 });
        }
        if (delay && delay > 0) {
            await new Promise(resolve => setTimeout(resolve, delay));
        }
    }
    /**
     * Scrape a webpage
     */
    async scrape(params) {
        // Handle both old (string) and new (params) API
        const opts = typeof params === 'string'
            ? { url: params }
            : params;
        const { page, cleanup } = await this.getPage(opts.sessionId);
        try {
            await page.goto(opts.url, { waitUntil: 'networkidle2' });
            await this.waitForPage(page, opts.delay, opts.waitFor);
            const title = await page.title();
            const html = await page.content();
            let content = html;
            let markdown;
            // Format conversion
            if (opts.format === 'text') {
                content = await page.evaluate(() => document.body.innerText);
            }
            else if (opts.format === 'markdown') {
                // Basic HTML to markdown conversion
                content = await page.evaluate(() => document.body.innerText);
                markdown = content; // Placeholder - use turndown for proper conversion
            }
            else if (opts.format === 'readability') {
                // Readability extraction
                content = await page.evaluate(() => {
                    // Simple extraction of main content
                    const article = document.querySelector('article') ||
                        document.querySelector('main') ||
                        document.querySelector('.content') ||
                        document.body;
                    return article?.innerText || '';
                });
            }
            // Extract metadata
            const metadata = await page.evaluate(() => {
                const metas = {};
                document.querySelectorAll('meta').forEach(meta => {
                    const name = meta.getAttribute('name') || meta.getAttribute('property');
                    const content = meta.getAttribute('content');
                    if (name && content) {
                        metas[name] = content;
                    }
                });
                return metas;
            });
            return {
                title,
                content,
                url: opts.url,
                html: opts.format !== 'html' ? undefined : html,
                markdown,
                metadata
            };
        }
        finally {
            await cleanup();
        }
    }
    /**
     * Take a screenshot
     */
    async screenshot(params, fullPage) {
        // Handle both old (string) and new (params) API
        const opts = typeof params === 'string'
            ? { url: params, fullPage: fullPage ?? true }
            : params;
        const { page, cleanup } = await this.getPage(opts.sessionId);
        try {
            await page.goto(opts.url, { waitUntil: 'networkidle2' });
            if (opts.delay && opts.delay > 0) {
                await new Promise(resolve => setTimeout(resolve, opts.delay));
            }
            const viewport = page.viewport();
            const data = await page.screenshot({
                fullPage: opts.fullPage ?? true,
                encoding: 'binary',
                type: opts.format || 'png',
                quality: opts.format === 'jpeg' ? opts.quality : undefined
            });
            // Return old format for backwards compatibility if called with string
            if (typeof params === 'string') {
                return data;
            }
            return {
                data,
                format: opts.format || 'png',
                width: viewport?.width || 1920,
                height: viewport?.height || 1080
            };
        }
        finally {
            await cleanup();
        }
    }
    /**
     * Generate PDF
     */
    async pdf(params) {
        // Handle both old (string) and new (params) API
        const opts = typeof params === 'string'
            ? { url: params }
            : params;
        const { page, cleanup } = await this.getPage(opts.sessionId);
        try {
            await page.goto(opts.url, { waitUntil: 'networkidle2' });
            const pdfBuffer = await page.pdf({
                format: opts.format || 'A4',
                landscape: opts.landscape,
                printBackground: opts.printBackground ?? true,
                margin: opts.margin
            });
            const data = Buffer.from(pdfBuffer);
            // Return old format for backwards compatibility if called with string
            if (typeof params === 'string') {
                return data;
            }
            return {
                data,
                pageCount: 1 // PDF page count would require parsing the PDF
            };
        }
        finally {
            await cleanup();
        }
    }
}
exports.QuickActionsService = QuickActionsService;
