// @testmuai/browser-cloud/dist/testmu-cloud/services/tunnel-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TunnelService = void 0;
const node_tunnel_1 = require("@lambdatest/node-tunnel");
class TunnelService {
    constructor() {
        this.tunnelInstance = null;
    }
    async start(config) {
        if (this.getStatus()) {
            console.log('Tunnel is already running.');
            return;
        }
        console.log('Starting LambdaTest Tunnel...');
        this.tunnelInstance = new node_tunnel_1.Tunnel();
        // Convert config to Tunnel Arguments
        // https://www.lambdatest.com/support/docs/tunnel-modifiers/
        const tunnelArgs = {
            user: config.user,
            key: config.key,
            tunnelName: config.tunnelName || `browser-cloud_Tunnel_${Date.now()}`,
            v: true, // Verbose logging
        };
        if (config.proxyHost) {
            tunnelArgs.proxyHost = config.proxyHost;
            tunnelArgs.proxyPort = config.proxyPort;
            if (config.proxyUser) {
                tunnelArgs.proxyUser = config.proxyUser;
                tunnelArgs.proxyPass = config.proxyPass;
            }
        }
        // Log file for debugging
        if (config.logFile) {
            tunnelArgs.logFile = config.logFile;
        }
        return new Promise((resolve, reject) => {
            this.tunnelInstance?.start(tunnelArgs, (error) => {
                if (error) {
                    console.error('Failed to start tunnel:', error);
                    reject(error);
                }
                else {
                    console.log('Tunnel Started Successfully!');
                    resolve();
                }
            });
        });
    }
    async stop() {
        if (!this.tunnelInstance || !this.getStatus()) {
            return;
        }
        console.log('Stopping LambdaTest Tunnel...');
        return new Promise((resolve) => {
            this.tunnelInstance?.stop(() => {
                console.log('Tunnel Stopped.');
                this.tunnelInstance = null;
                resolve();
            });
        });
    }
    getStatus() {
        if (!this.tunnelInstance)
            return false;
        return this.tunnelInstance.isRunning();
    }
}
exports.TunnelService = TunnelService;
