// @testmuai/browser-cloud/dist/testmu-cloud/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventsService = exports.ContextService = exports.ComputerService = exports.TunnelService = exports.CaptchaService = exports.ProfileService = exports.CredentialService = exports.ExtensionService = exports.FileService = exports.QuickActionsService = exports.SeleniumAdapter = exports.PlaywrightAdapter = exports.PuppeteerAdapter = exports.SessionManager = exports.Browser = void 0;
const session_manager_js_1 = require("./session-manager.js");
// Adapters
const puppeteer_js_1 = require("./adapters/puppeteer.js");
const playwright_js_1 = require("./adapters/playwright.js");
const selenium_js_1 = require("./adapters/selenium.js");
// Services
const quick_actions_js_1 = require("./services/quick-actions.js");
const file_service_js_1 = require("./services/file-service.js");
const extension_service_js_1 = require("./services/extension-service.js");
const credential_service_js_1 = require("./services/credential-service.js");
const profile_service_js_1 = require("./profile-service.js");
const captcha_service_js_1 = require("./services/captcha-service.js");
const tunnel_service_js_1 = require("./services/tunnel-service.js");
const computer_service_js_1 = require("./services/computer-service.js");
const context_service_js_1 = require("./services/context-service.js");
const events_service_js_1 = require("./services/events-service.js");
/**
 *
 * Provides an AI Browser Automation SDK for TestMu AI Browser Cloud.
 * Features include:
 * - Session management with full TestMu AI Browser Cloud API parity
 * - AI agent computer actions (mouse, keyboard, screenshots)
 * - Browser context management (cookies, localStorage, sessionStorage)
 * - Captcha solving integration
 * - File and extension management
 * - Profile persistence
 *
 * @example
 * ```typescript
 * import { Browser } from '@testmuai/browser-cloud';
 *
 * const browser = new Browser();
 * const session = await browser.sessions.create({ local: true });
 * const page = await browser.puppeteer.connect(session);
 *
 * // Use AI agent actions
 * await browser.sessions.computer(session.id, page, { action: 'click', coordinate: [100, 200] });
 *
 * await browser.sessions.release(session.id);
 * ```
 */
class Browser {
    constructor() {
        // ================== Sessions API (Full TestMu AI Browser Cloud Parity) ==================
        this.sessions = {
            /**
             * Create a new browser session
             */
            create: async (config = {}) => {
                return await this.sessionManager.createSession(config);
            },
            /**
             * List all active sessions
             */
            list: () => {
                return this.sessionManager.listSessions();
            },
            /**
             * Retrieve a specific session by ID
             */
            retrieve: (sessionId) => {
                return this.sessionManager.getSession(sessionId);
            },
            /**
             * Release a specific session
             */
            release: async (sessionId) => {
                return await this.sessionManager.releaseSession(sessionId);
            },
            /**
             * Release all active sessions
             */
            releaseAll: async () => {
                return await this.sessionManager.releaseAll();
            },
            /**
             * Execute computer actions (mouse, keyboard, screenshot)
             * This is the key API for AI agents
             */
            computer: async (sessionId, page, params) => {
                return await this.computer.execute(page, params);
            },
            /**
             * Get session context (cookies, localStorage, sessionStorage)
             */
            context: async (sessionId, page) => {
                return await this.context.getContext(page);
            },
            /**
             * Get recorded session events (RRWeb format)
             */
            events: (sessionId) => {
                return this.sessionManager.getSessionEvents(sessionId);
            },
            /**
             * Get live session details (pages, tabs, state)
             */
            liveDetails: async (sessionId) => {
                return await this.sessionManager.getLiveDetails(sessionId);
            },
            // Session-scoped file operations
            files: {
                list: async (sessionId) => {
                    return this.files.listSessionFiles(sessionId);
                },
                upload: async (sessionId, file, filename) => {
                    return this.files.uploadToSession(sessionId, file, filename);
                },
                download: async (sessionId, path) => {
                    return this.files.downloadFromSession(sessionId, path);
                },
                downloadArchive: async (sessionId) => {
                    return this.files.downloadSessionArchive(sessionId);
                },
                delete: async (sessionId, path) => {
                    return this.files.deleteFromSession(sessionId, path);
                },
                deleteAll: async (sessionId) => {
                    return this.files.deleteAllSessionFiles(sessionId);
                }
            },
            // Session-scoped captcha operations
            captchas: {
                solveImage: async (sessionId, params) => {
                    return this.captcha.solveImage(sessionId, params);
                },
                status: async (sessionId) => {
                    return this.captcha.status(sessionId);
                }
            }
        };
        this.sessionManager = new session_manager_js_1.SessionManager();
        // Adapters
        this.puppeteer = new puppeteer_js_1.PuppeteerAdapter();
        this.playwright = new playwright_js_1.PlaywrightAdapter();
        this.selenium = new selenium_js_1.SeleniumAdapter();
        // Services
        this.quick = new quick_actions_js_1.QuickActionsService();
        this.files = new file_service_js_1.FileService();
        this.extensions = new extension_service_js_1.ExtensionService();
        this.credentials = new credential_service_js_1.CredentialService();
        this.profiles = new profile_service_js_1.ProfileService();
        this.captcha = new captcha_service_js_1.CaptchaService();
        this.tunnel = new tunnel_service_js_1.TunnelService();
        // New TestMu AI Browser Cloud Parity Services
        this.computer = new computer_service_js_1.ComputerService();
        this.context = new context_service_js_1.ContextService();
        this.events = new events_service_js_1.EventsService();
        // Pass services to SessionManager for automatic handling
        this.sessionManager.setTunnelService(this.tunnel);
        this.sessionManager.setExtensionService(this.extensions);
    }
    // ================== Quick Action Aliases (TestMu AI Browser Cloud SDK style) ==================
    /**
     * Scrape a webpage - shorthand for quick.scrape()
     */
    async scrape(params) {
        return this.quick.scrape(params);
    }
    /**
     * Take a screenshot - shorthand for quick.screenshot()
     */
    async screenshot(params, fullPage) {
        return this.quick.screenshot(params, fullPage);
    }
    /**
     * Generate PDF - shorthand for quick.pdf()
     */
    async pdf(params) {
        return this.quick.pdf(params);
    }
}
exports.Browser = Browser;
// ================== Export Types ==================
__exportStar(require("./types.js"), exports);
// ================== Export Services ==================
var session_manager_js_2 = require("./session-manager.js");
Object.defineProperty(exports, "SessionManager", { enumerable: true, get: function () { return session_manager_js_2.SessionManager; } });
var puppeteer_js_2 = require("./adapters/puppeteer.js");
Object.defineProperty(exports, "PuppeteerAdapter", { enumerable: true, get: function () { return puppeteer_js_2.PuppeteerAdapter; } });
var playwright_js_2 = require("./adapters/playwright.js");
Object.defineProperty(exports, "PlaywrightAdapter", { enumerable: true, get: function () { return playwright_js_2.PlaywrightAdapter; } });
var selenium_js_2 = require("./adapters/selenium.js");
Object.defineProperty(exports, "SeleniumAdapter", { enumerable: true, get: function () { return selenium_js_2.SeleniumAdapter; } });
var quick_actions_js_2 = require("./services/quick-actions.js");
Object.defineProperty(exports, "QuickActionsService", { enumerable: true, get: function () { return quick_actions_js_2.QuickActionsService; } });
var file_service_js_2 = require("./services/file-service.js");
Object.defineProperty(exports, "FileService", { enumerable: true, get: function () { return file_service_js_2.FileService; } });
var extension_service_js_2 = require("./services/extension-service.js");
Object.defineProperty(exports, "ExtensionService", { enumerable: true, get: function () { return extension_service_js_2.ExtensionService; } });
var credential_service_js_2 = require("./services/credential-service.js");
Object.defineProperty(exports, "CredentialService", { enumerable: true, get: function () { return credential_service_js_2.CredentialService; } });
var profile_service_js_2 = require("./profile-service.js");
Object.defineProperty(exports, "ProfileService", { enumerable: true, get: function () { return profile_service_js_2.ProfileService; } });
var captcha_service_js_2 = require("./services/captcha-service.js");
Object.defineProperty(exports, "CaptchaService", { enumerable: true, get: function () { return captcha_service_js_2.CaptchaService; } });
var tunnel_service_js_2 = require("./services/tunnel-service.js");
Object.defineProperty(exports, "TunnelService", { enumerable: true, get: function () { return tunnel_service_js_2.TunnelService; } });
var computer_service_js_2 = require("./services/computer-service.js");
Object.defineProperty(exports, "ComputerService", { enumerable: true, get: function () { return computer_service_js_2.ComputerService; } });
var context_service_js_2 = require("./services/context-service.js");
Object.defineProperty(exports, "ContextService", { enumerable: true, get: function () { return context_service_js_2.ContextService; } });
var events_service_js_2 = require("./services/events-service.js");
Object.defineProperty(exports, "EventsService", { enumerable: true, get: function () { return events_service_js_2.EventsService; } });
