// @testmuai/browser-cloud/dist/testmu-cloud/profile-service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProfileService = void 0;
const fs = __importStar(require("fs-extra"));
const path = __importStar(require("path"));
const context_service_js_1 = require("./services/context-service.js");
/**
 * ProfileService - Browser Profile Management for Deployed SDK
 *
 * Manages persistent browser profiles for session state.
 * Profiles store cookies, localStorage, and sessionStorage.
 *
 * Storage location (in order of priority):
 * 1. Configured via setConfig({ profilesDir: '/path' })
 * 2. Environment variable: TESTMU_PROFILES_DIR
 * 3. Default: {user's project}/.profiles/
 *
 * Usage:
 * ```typescript
 * const client = new Browser();
 *
 * // Optional: Configure custom profiles directory
 * client.profiles.setConfig({ profilesDir: './my-profiles' });
 *
 * // Save browser state to profile
 * await client.profiles.saveProfile('logged-in-user', page);
 *
 * // Load profile into new session
 * await client.profiles.loadProfile('logged-in-user', page);
 * ```
 */
class ProfileService {
    constructor(config) {
        this.initialized = false;
        this.contextService = new context_service_js_1.ContextService();
        this.config = {
            profilesDir: config?.profilesDir || process.env.TESTMU_PROFILES_DIR || undefined,
            verbose: config?.verbose ?? true
        };
    }
    /**
     * Configure the profile service
     * Call this before using profiles if you need custom settings
     */
    setConfig(config) {
        this.config = { ...this.config, ...config };
        this.initialized = false; // Reset to reinitialize with new config
    }
    /**
     * Get the profiles directory path
     */
    getProfilesDir() {
        return this.config.profilesDir || path.join(process.cwd(), '.profiles');
    }
    /**
     * Ensure profiles directory exists (lazy initialization)
     */
    async ensureDir() {
        if (!this.initialized) {
            const dir = this.getProfilesDir();
            await fs.ensureDir(dir);
            this.initialized = true;
            if (this.config.verbose) {
                console.log(`[ProfileService] Using profiles directory: ${dir}`);
            }
        }
    }
    /**
     * Get file path for a profile
     */
    getProfilePath(id) {
        return path.join(this.getProfilesDir(), `${id}.json`);
    }
    /**
     * Log message if verbose mode is enabled
     */
    log(message) {
        if (this.config.verbose) {
            console.log(`[ProfileService] ${message}`);
        }
    }
    // ================== Core Profile Operations ==================
    /**
     * Load a profile into a page (inject cookies, localStorage, sessionStorage)
     *
     * @param id - Profile ID
     * @param page - Puppeteer page (can be from LambdaTest cloud session)
     * @param options - Optional settings
     */
    async loadProfile(id, page, options) {
        await this.ensureDir();
        const filePath = this.getProfilePath(id);
        if (!await fs.pathExists(filePath)) {
            this.log(`Profile '${id}' not found at ${filePath}`);
            return false;
        }
        try {
            const profile = await fs.readJson(filePath);
            // Navigate first if specified (needed for localStorage/sessionStorage)
            if (options?.navigateFirst) {
                await page.goto(options.navigateFirst, { waitUntil: 'domcontentloaded' });
            }
            if (options?.cookiesOnly) {
                // Only load cookies
                await this.contextService.setCookies(page, profile.cookies);
                this.log(`Loaded cookies from profile '${id}' (${profile.cookies.length} cookies)`);
            }
            else {
                // Get the current page origin to use as key for localStorage/sessionStorage
                const currentOrigin = await page.evaluate(() => window.location.origin);
                // Load full context - use current origin so setContext can match it
                const context = {
                    cookies: profile.cookies,
                    localStorage: { [currentOrigin]: profile.localStorage },
                    sessionStorage: { [currentOrigin]: profile.sessionStorage }
                };
                await this.contextService.setContext(page, context);
                this.log(`Loaded profile '${id}' (${profile.cookies.length} cookies, localStorage, sessionStorage)`);
            }
            return true;
        }
        catch (error) {
            this.log(`Error loading profile '${id}': ${error.message}`);
            return false;
        }
    }
    /**
     * Save page state to a profile
     *
     * @param id - Profile ID (will create or update)
     * @param page - Puppeteer page (can be from LambdaTest cloud session)
     * @param options - Optional settings
     */
    async saveProfile(id, page, options) {
        await this.ensureDir();
        // Get context from page
        const context = await this.contextService.getContext(page);
        // Flatten localStorage/sessionStorage for storage
        const localStorage = Object.values(context.localStorage || {})[0] || {};
        const sessionStorage = Object.values(context.sessionStorage || {})[0] || {};
        const filePath = this.getProfilePath(id);
        const now = new Date().toISOString();
        // Build profile data
        const profile = {
            id,
            name: options?.name,
            description: options?.description,
            cookies: context.cookies || [],
            localStorage,
            sessionStorage,
            createdAt: now,
            updatedAt: now,
            metadata: options?.metadata
        };
        // Preserve createdAt from existing profile
        if (await fs.pathExists(filePath)) {
            try {
                const existing = await fs.readJson(filePath);
                profile.createdAt = existing.createdAt || now;
            }
            catch {
                // Ignore read errors, use new createdAt
            }
        }
        await fs.writeJson(filePath, profile, { spaces: 2 });
        this.log(`Saved profile '${id}' (${profile.cookies.length} cookies)`);
        return profile;
    }
    // ================== Profile Management ==================
    /**
     * Create an empty profile
     */
    async create(id, options) {
        await this.ensureDir();
        const filePath = this.getProfilePath(id);
        const now = new Date().toISOString();
        const profile = {
            id,
            name: options?.name,
            description: options?.description,
            cookies: [],
            localStorage: {},
            sessionStorage: {},
            createdAt: now,
            updatedAt: now
        };
        await fs.writeJson(filePath, profile, { spaces: 2 });
        this.log(`Created empty profile '${id}'`);
        return profile;
    }
    /**
     * Get a profile by ID
     */
    async get(id) {
        await this.ensureDir();
        const filePath = this.getProfilePath(id);
        if (!await fs.pathExists(filePath)) {
            return null;
        }
        return fs.readJson(filePath);
    }
    /**
     * List all profiles
     */
    async list() {
        await this.ensureDir();
        const dir = this.getProfilesDir();
        if (!await fs.pathExists(dir)) {
            return [];
        }
        const files = await fs.readdir(dir);
        const profiles = [];
        for (const file of files) {
            if (file.endsWith('.json')) {
                try {
                    const profile = await fs.readJson(path.join(dir, file));
                    profiles.push(profile);
                }
                catch {
                    // Skip invalid files
                }
            }
        }
        return profiles;
    }
    /**
     * Check if a profile exists
     */
    async exists(id) {
        const filePath = this.getProfilePath(id);
        return fs.pathExists(filePath);
    }
    /**
     * Delete a profile
     */
    async delete(id) {
        const filePath = this.getProfilePath(id);
        if (await fs.pathExists(filePath)) {
            await fs.remove(filePath);
            this.log(`Deleted profile '${id}'`);
            return true;
        }
        return false;
    }
    /**
     * Delete all profiles
     */
    async deleteAll() {
        await this.ensureDir();
        const profiles = await this.list();
        for (const profile of profiles) {
            await this.delete(profile.id);
        }
        this.log(`Deleted ${profiles.length} profiles`);
        return profiles.length;
    }
    /**
     * Duplicate a profile
     */
    async duplicate(sourceId, targetId) {
        const source = await this.get(sourceId);
        if (!source) {
            throw new Error(`Source profile '${sourceId}' not found`);
        }
        const now = new Date().toISOString();
        const duplicate = {
            ...source,
            id: targetId,
            name: source.name ? `${source.name} (copy)` : undefined,
            createdAt: now,
            updatedAt: now
        };
        await fs.writeJson(this.getProfilePath(targetId), duplicate, { spaces: 2 });
        this.log(`Duplicated profile '${sourceId}' to '${targetId}'`);
        return duplicate;
    }
    // ================== Utility Methods ==================
    /**
     * Export a profile to a portable format
     */
    async export(id) {
        const profile = await this.get(id);
        if (!profile) {
            throw new Error(`Profile '${id}' not found`);
        }
        return JSON.stringify(profile, null, 2);
    }
    /**
     * Import a profile from JSON string
     */
    async import(json, options) {
        await this.ensureDir();
        const profile = JSON.parse(json);
        if (!profile.id) {
            throw new Error('Invalid profile: missing id');
        }
        const exists = await this.exists(profile.id);
        if (exists && !options?.overwrite) {
            throw new Error(`Profile '${profile.id}' already exists. Use overwrite: true to replace.`);
        }
        profile.updatedAt = new Date().toISOString();
        await fs.writeJson(this.getProfilePath(profile.id), profile, { spaces: 2 });
        this.log(`Imported profile '${profile.id}'`);
        return profile;
    }
    // ================== Legacy Methods (for backwards compatibility) ==================
    /** @deprecated Use get() instead */
    async getProfile(id) {
        return this.get(id);
    }
    /** @deprecated Use list() instead */
    async listProfiles() {
        return this.list();
    }
    /** @deprecated Use saveProfile() instead */
    async createFromSession(sessionId, page) {
        const profileId = `profile_${sessionId}_${Date.now()}`;
        await this.saveProfile(profileId, page);
        return { id: profileId, status: 'READY' };
    }
}
exports.ProfileService = ProfileService;
