// @testmuai/browser-cloud/dist/testmu-cloud/session-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionManager = void 0;
const local_browser_service_js_1 = require("./services/local-browser-service.js");
const events_service_js_1 = require("./services/events-service.js");
const stealth_utils_js_1 = require("./stealth-utils.js");
class SessionManager {
    constructor() {
        this.tunnelService = null;
        this.extensionService = null;
        this.eventsService = new events_service_js_1.EventsService();
    }
    setTunnelService(service) {
        this.tunnelService = service;
    }
    setExtensionService(service) {
        this.extensionService = service;
    }
    async createSession(config) {
        // Generate unique session ID with timestamp + random suffix
        const randomSuffix = Math.random().toString(36).substring(2, 8);
        const sessionId = config.sessionId || `session_${Date.now()}_${randomSuffix}`;
        const createdAt = new Date().toISOString();
        // Auto-pick a random user-agent if stealth is configured and no explicit UA is set
        let resolvedUserAgent = config.userAgent;
        if (config.stealthConfig && config.stealthConfig.randomizeUserAgent !== false && !config.userAgent) {
            resolvedUserAgent = (0, stealth_utils_js_1.getRandomUserAgent)();
            console.log(`SessionManager: Auto-selected stealth user-agent: ${resolvedUserAgent.substring(0, 50)}...`);
        }
        if (config.customWebSocketUrl) {
            // Bring Your Own Browser (BYOB)
            console.log(`Connecting to custom WebSocket URL: ${config.customWebSocketUrl}`);
            const session = {
                id: sessionId,
                websocketUrl: config.customWebSocketUrl,
                debugUrl: 'about:blank',
                config,
                status: 'live',
                createdAt,
                timeout: config.timeout || 300000,
                dimensions: config.dimensions || { width: 1920, height: 1080 }
            };
            SessionStore.start(session, null);
            this.eventsService.startRecording(sessionId);
            return session;
        }
        if (config.local) {
            // Launch local browser using chrome-launcher (Auto-discovery)
            console.log("Launching local browser (auto-discovery)...");
            const localService = new local_browser_service_js_1.LocalBrowserService();
            const { websocketUrl, kill } = await localService.launch();
            const session = {
                id: sessionId,
                websocketUrl: websocketUrl,
                debugUrl: "http://localhost:9222",
                sessionViewerUrl: "http://localhost:9222",
                config,
                status: 'live',
                createdAt,
                timeout: config.timeout || 300000,
                dimensions: config.dimensions || { width: 1920, height: 1080 },
                headless: config.headless,
                stealthConfig: config.stealthConfig,
                userAgent: resolvedUserAgent
            };
            SessionStore.start(session, null, kill);
            this.eventsService.startRecording(sessionId);
            return session;
        }
        else {
            // LambdaTest
            const username = process.env.LT_USERNAME || 'generic_user';
            const accessKey = process.env.LT_ACCESS_KEY || 'generic_key';
            // Map adapter to LambdaTest plugin name and WebSocket endpoint
            const adapter = config.adapter || 'puppeteer';
            const adapterPluginMap = {
                'puppeteer': 'node-js-puppeteer',
                'playwright': 'node-js-playwright',
                'selenium': 'node-js-selenium'
            };
            const adapterEndpointMap = {
                'puppeteer': 'puppeteer',
                'playwright': 'playwright',
                'selenium': 'selenium'
            };
            const plugin = adapterPluginMap[adapter];
            const wsEndpointPath = adapterEndpointMap[adapter];
            // Extract user's LT:Options if nested inside lambdatestOptions
            const userLtOptions = config.lambdatestOptions?.['LT:Options'] || {};
            const { 'LT:Options': _removed, ...restLambdatestOptions } = config.lambdatestOptions || {};
            const ltOptions = {
                platformName: 'Windows 10',
                project: 'browser-cloud',
                w3c: true,
                plugin,
                ...restLambdatestOptions,
                ...userLtOptions
            };
            // Handle proxy configuration
            if (config.proxy || config.proxyUrl) {
                ltOptions.proxy = config.proxy || config.proxyUrl;
            }
            // Handle tunnel
            if (config.tunnel) {
                ltOptions.tunnel = true;
                if (config.tunnelName) {
                    ltOptions.tunnelName = config.tunnelName;
                }
                else if (this.tunnelService) {
                    const managedName = `browser-cloud_Managed_${Date.now()}`;
                    if (!this.tunnelService.getStatus()) {
                        await this.tunnelService.start({
                            user: username,
                            key: accessKey,
                            tunnelName: managedName
                        });
                    }
                    ltOptions.tunnelName = managedName;
                }
                else {
                    console.warn("Tunnel requested but TunnelService not initialized.");
                }
            }
            // Handle geolocation
            if (config.geoLocation) {
                ltOptions.geoLocation = config.geoLocation;
            }
            // Handle region
            if (config.region) {
                ltOptions.region = config.region;
            }
            // Handle resolution/dimensions
            if (config.dimensions) {
                ltOptions.resolution = `${config.dimensions.width}x${config.dimensions.height}`;
            }
            // Handle headless
            if (config.headless !== undefined) {
                ltOptions.headless = config.headless;
            }
            // Handle extensions via extensionIds config
            if (config.extensionIds && config.extensionIds.length > 0 && this.extensionService) {
                const cloudUrls = await this.extensionService.getCloudUrls(config.extensionIds);
                if (cloudUrls.length > 0) {
                    // Merge with any existing lambda:loadExtension URLs
                    const existing = ltOptions['lambda:loadExtension'] || [];
                    ltOptions['lambda:loadExtension'] = [...existing, ...cloudUrls];
                    console.log(`Loading ${cloudUrls.length} extension(s) into session`);
                }
            }
            const capabilities = {
                browserName: 'Chrome',
                browserVersion: 'latest',
                'LT:Options': ltOptions
            };
            // LambdaTest requires authentication in the URL
            // Use correct endpoint based on adapter (puppeteer/playwright/selenium)
            const wsEndpoint = `wss://${username}:${accessKey}@cdp.lambdatest.com/${wsEndpointPath}?capabilities=${encodeURIComponent(JSON.stringify(capabilities))}`;
            const session = {
                id: sessionId,
                websocketUrl: wsEndpoint,
                debugUrl: "https://automation.lambdatest.com/logs/",
                sessionViewerUrl: "https://automation.lambdatest.com/logs/",
                config,
                status: 'live',
                createdAt,
                timeout: config.timeout || 300000,
                dimensions: config.dimensions || { width: 1920, height: 1080 },
                headless: config.headless,
                isSelenium: config.isSelenium,
                persistProfile: config.persistProfile,
                profileId: config.profileId,
                region: config.region,
                solveCaptcha: config.solveCaptcha,
                stealthConfig: config.stealthConfig,
                userAgent: resolvedUserAgent
            };
            SessionStore.start(session, null);
            this.eventsService.startRecording(sessionId);
            return session;
        }
    }
    listSessions() {
        return SessionStore.list();
    }
    getSession(id) {
        return SessionStore.get(id)?.session;
    }
    async releaseSession(id) {
        const entry = SessionStore.get(id);
        if (entry) {
            console.log(`Closing session ${id}`);
            // Stop recording
            this.eventsService.stopRecording(id);
            // Update session status
            entry.session.status = 'released';
            if (entry.cleanup) {
                await entry.cleanup();
            }
            else if (entry.browser) {
                await entry.browser.close();
            }
            SessionStore.delete(id);
            return { success: true, message: `Session ${id} released` };
        }
        return { success: false, message: `Session ${id} not found` };
    }
    /**
     * Release all active sessions
     */
    async releaseAll() {
        const sessions = SessionStore.list();
        let releasedCount = 0;
        for (const session of sessions) {
            try {
                await this.releaseSession(session.id);
                releasedCount++;
            }
            catch (e) {
                console.error(`Error releasing session ${session.id}:`, e);
            }
        }
        return {
            success: true,
            message: `Released ${releasedCount} sessions`
        };
    }
    /**
     * Get recorded events for a session (RRWeb format)
     */
    getSessionEvents(id) {
        return this.eventsService.getEvents(id);
    }
    /**
     * Get event count for a session
     */
    getEventCount(id) {
        return this.eventsService.getEventCount(id);
    }
    /**
     * Get live details for a session (pages, tabs, state)
     */
    async getLiveDetails(id) {
        const entry = SessionStore.get(id);
        if (!entry || !entry.session) {
            return null;
        }
        const session = entry.session;
        // For now, return basic info
        // Full implementation would track open pages via CDP
        const pages = [{
                id: 'page_0',
                url: 'about:blank',
                title: 'Browser Tab',
                favicon: null,
                sessionViewerUrl: session.sessionViewerUrl || session.debugUrl,
                sessionViewerFullscreenUrl: session.sessionViewerUrl || session.debugUrl
            }];
        return {
            pages,
            wsUrl: session.websocketUrl,
            sessionViewerUrl: session.sessionViewerUrl || session.debugUrl,
            sessionViewerFullscreenUrl: session.sessionViewerUrl || session.debugUrl
        };
    }
}
exports.SessionManager = SessionManager;
// Enhanced Session Store to support list/retrieve parity
class SessionRepository {
    constructor() {
        this.sessions = new Map();
    }
    start(session, browser, cleanup) {
        this.sessions.set(session.id, { session, browser, cleanup });
    }
    get(id) {
        return this.sessions.get(id);
    }
    delete(id) {
        this.sessions.delete(id);
    }
    list() {
        return Array.from(this.sessions.values()).map(data => data.session);
    }
    getAll() {
        return Array.from(this.sessions.values());
    }
    count() {
        return this.sessions.size;
    }
}
const SessionStore = new SessionRepository();
