// @testmuai/browser-cloud/dist/testmu-cloud/adapters/puppeteer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PuppeteerAdapter = void 0;
const puppeteer_core_1 = __importDefault(require("puppeteer-core"));
const puppeteer_extra_1 = __importDefault(require("puppeteer-extra"));
const puppeteer_extra_plugin_stealth_1 = __importDefault(require("puppeteer-extra-plugin-stealth"));
const profile_service_js_1 = require("../profile-service.js");
const stealth_utils_js_1 = require("../stealth-utils.js");
puppeteer_extra_1.default.use((0, puppeteer_extra_plugin_stealth_1.default)());
class PuppeteerAdapter {
    constructor() {
        this.profileService = new profile_service_js_1.ProfileService();
    }
    async connect(session) {
        console.log(`Adapter: Connecting to session ${session.id} via Puppeteer...`);
        const stealthConfig = session.stealthConfig || session.config.stealthConfig;
        const skipFingerprint = stealthConfig?.skipFingerprintInjection === true;
        let browser;
        try {
            if (skipFingerprint) {
                // Raw puppeteer without stealth plugin
                console.log('Adapter: Skipping fingerprint injection (raw puppeteer)');
                browser = await puppeteer_core_1.default.connect({
                    browserWSEndpoint: session.websocketUrl,
                });
            }
            else {
                // Apply Stealth by using puppeteer-extra
                browser = await puppeteer_extra_1.default.connect({
                    browserWSEndpoint: session.websocketUrl,
                });
            }
        }
        catch (e) {
            console.error("Puppeteer Adapter: Connection failed", e);
            throw e;
        }
        // Fetch LambdaTest dashboard URL (best effort, cloud sessions only)
        if (!session.config.local && !session.config.customWebSocketUrl) {
            await this.fetchDashboardUrl(session);
        }
        // Apply stealth UA / viewport if stealth is configured
        if (stealthConfig) {
            const pages = await browser.pages();
            const page = pages.length > 0 ? pages[0] : await browser.newPage();
            // Random user-agent
            if (stealthConfig.randomizeUserAgent !== false) {
                const ua = session.userAgent || (0, stealth_utils_js_1.getRandomUserAgent)();
                await page.setUserAgent(ua);
                console.log(`Adapter: Set stealth user-agent: ${ua.substring(0, 50)}...`);
            }
            // Random viewport
            if (stealthConfig.randomizeViewport !== false) {
                const dims = session.dimensions || { width: 1920, height: 1080 };
                const vp = (0, stealth_utils_js_1.getRandomizedViewport)(dims.width, dims.height);
                await page.setViewport(vp);
                console.log(`Adapter: Set stealth viewport: ${vp.width}x${vp.height}`);
            }
            // Humanize interactions
            if (stealthConfig.humanizeInteractions) {
                this.applyHumanize(page);
                // Apply to future pages
                browser.on('targetcreated', async (target) => {
                    if (target.type() === 'page') {
                        const newPage = await target.page();
                        if (newPage) {
                            this.applyHumanize(newPage);
                            if (stealthConfig.randomizeUserAgent !== false) {
                                const ua = session.userAgent || (0, stealth_utils_js_1.getRandomUserAgent)();
                                await newPage.setUserAgent(ua);
                            }
                            if (stealthConfig.randomizeViewport !== false) {
                                const dims = session.dimensions || { width: 1920, height: 1080 };
                                const vp = (0, stealth_utils_js_1.getRandomizedViewport)(dims.width, dims.height);
                                await newPage.setViewport(vp);
                            }
                        }
                    }
                });
                console.log('Adapter: Humanized interactions enabled');
            }
        }
        // Apply Persistence
        if (session.config.profileId) {
            const pages = await browser.pages();
            const page = pages.length > 0 ? pages[0] : await browser.newPage();
            console.log(`Adapter: Loading profile ${session.config.profileId}`);
            await this.profileService.loadProfile(session.config.profileId, page);
            // Wrap browser.close to save profile
            const originalClose = browser.close.bind(browser);
            browser.close = async () => {
                console.log(`Adapter: Saving profile ${session.config.profileId} on close...`);
                try {
                    const currentPages = await browser.pages();
                    if (currentPages.length > 0) {
                        await this.profileService.saveProfile(session.config.profileId, currentPages[0]);
                    }
                    else {
                        console.warn("Adapter: No pages open to save profile.");
                    }
                }
                catch (err) {
                    console.error("Adapter: Failed to save profile on close", err);
                }
                return originalClose();
            };
        }
        return browser;
    }
    async fetchDashboardUrl(session) {
        const username = process.env.LT_USERNAME;
        const accessKey = process.env.LT_ACCESS_KEY;
        if (!username || !accessKey)
            return;
        try {
            const https = await Promise.resolve().then(() => __importStar(require('https')));
            const auth = Buffer.from(`${username}:${accessKey}`).toString('base64');
            const data = await new Promise((resolve, reject) => {
                const req = https.request({
                    hostname: 'api.lambdatest.com',
                    path: '/automation/api/v1/sessions?limit=1',
                    method: 'GET',
                    headers: {
                        'Authorization': `Basic ${auth}`,
                        'Accept': 'application/json'
                    }
                }, (res) => {
                    let body = '';
                    res.on('data', (chunk) => body += chunk);
                    res.on('end', () => resolve(body));
                });
                req.on('error', reject);
                req.setTimeout(10000, () => { req.destroy(); reject(new Error('timeout')); });
                req.end();
            });
            const json = JSON.parse(data);
            if (json.data && json.data.length > 0) {
                const entry = json.data[0];
                const testId = entry.test_id || entry.session_id;
                const buildId = entry.build_id;
                if (testId && buildId) {
                    const url = `https://automation.lambdatest.com/test?build=${buildId}&testID=${testId}`;
                    session.sessionViewerUrl = url;
                    console.log(`Puppeteer Adapter: LambdaTest Dashboard: ${url}`);
                }
            }
        }
        catch {
            // Best effort — don't fail the connection if we can't get the URL
        }
    }
    /**
     * Monkey-patch page.click() and page.type() to add random human-like delays.
     */
    applyHumanize(page) {
        const originalClick = page.click.bind(page);
        const originalType = page.type.bind(page);
        page.click = async (selector, options) => {
            // Random delay before click: 50-150ms
            await new Promise(r => setTimeout(r, 50 + Math.random() * 100));
            return originalClick(selector, options);
        };
        page.type = async (selector, text, options) => {
            // Type character by character with random delay: 30-130ms per char
            const delay = 30 + Math.random() * 100;
            return originalType(selector, text, { ...options, delay });
        };
    }
}
exports.PuppeteerAdapter = PuppeteerAdapter;
